iina/AtomicBoolSemaphore.swift

```diff
@@ -0,0 +1,39 @@
+//
+//  AtomicBoolSemaphore.swift
+//  iina
+//
+//  Created by low-batt on 9/21/22.
+//  Copyright © 2022 lhc. All rights reserved.
+//
+
+import Foundation
+
+/// An atomic boolean combined with a semaphore to allow waiting for the boolean to be set to `true`.
+///
+/// This class is expected to be used as follows:
+/// - This boolean is set to `false`, recreating the semaphore
+/// - A single thread waits for this boolean to be set to `true`
```

Comment:
Blocking a thread on a semaphore is typically not very polite on Darwin. Can we try rethinking this API and having work be queued instead?

Reply:
I too have been questioning my own decision to implement it this way. I don't think queuing would work. The correct way to avoid this is to toss the current thread based code in `AppDelegate.applicationShouldTerminate` in the trash and implement an event driven state machine to control shutdown. This is the offending code:

<details>
<summary><b>Thread Based Code:</b></summary>

```swift
    terminateQueue.async {
      // When a window is closed it stops playback. That involves sending a stop command to mpv.
      // That command executes asynchronously. MUST wait for that command to complete. If the quit
      // command is sent to mpv while the stop command is still executing the commands can
      // interfere and cause the playback position to not be saved in the watch later file. See
      // issue #3939 for details.
      for pc in PlayerCore.playerCores {
        pc.waitForStopToFinish(timeout: timeout)
      }
      // Now that playback has been stopped terminate the players.
      for pc in PlayerCore.playerCores {
        pc.terminateMPV()
      }
      // Player termination involves sending a quit command to mpv. That command executes
      // asynchronously. Wait for all the players to finish terminating before proceeding.
      for pc in PlayerCore.playerCores {
        pc.waitForTermination(timeout: timeout)
      }
      // Tell Cocoa to proceed with termination. This has to be done on the main thread.
      // The main dispatch queue MUST NOT be used to avoid a deadlock when quitting is
      // initiated by mpv in response to the user pressing "q". When that happens
      // MPVController submits a task to the main queue that calls terminate. That call
      // blocks the main dispatch queue.
      if #available(macOS 10.12, *) {
        RunLoop.main.perform(inModes: [.common, .eventTracking, .modalPanel]) {
          self.proceedWithTermination()
        }
      } else {
        RunLoop.main.perform(#selector(self.proceedWithTermination), target: self,
                             argument: nil, order: Int.min, modes: [.common])
      }
    }
```
</details>

The advantage of this thread style solution is that it is very clear what the shutdown process is doing as the sequence is right there in the code. An event driven solution is definitely better, but frequently not immediately obvious as to how the code works.

For an event driven system I envision something like the following...

`PlayerCore` would post notifications when the `stop` and `quit` commands finish. Early on `AppDelegate.applicationShouldTerminate` would add observers for those actions. When an event indicating `stop` as completed the handler in `AppDelegate` would have `PlayerCore` send a `quit` command. When an event indicating `quit` has completed the in `AppDelegate` would poll all the cores to see if they all have finished terminating and if so then proceed with termination. Something like that.

Unless you have second thoughts I'm thinking I should go ahead and change the implementation to use such an event driven approach.

Reply:
I take back my comments about waiting on a semaphore, the application is going to quit anyways and I don't really think it matters that this doesn't perform all that well. That said, I don't really like exposing this primitive as API, as we did with Lock (which I consider to be generally useful and desirable). Can we set up an ad-hoc dispatch group here and use it to control the forward progress of this thread?

Reply:
Originally I was thinking an event driven solution would require more changes and be more harder to understand. After thinking more on how that could be done I'm convinced event driven is the proper approach. That plus smarter use of dispatching to the main thread should be less complex than the current approach. The event driven approach eliminates the need for `AtomicBoolSemaphore`. By dispatching to the main thread I should be able to eliminate the need for `AtomicBool`.

Busy today, but should be able to post such a change later this week.